### ISSUE TITLE
Verification of Pre-trained GoogLeNet and ResNet50 weights

### ISSUE THREAD

**Issue by schrum2** [deeplearning4j/deeplearning4j#4139]
Has proper functioning of the GoogLeNet and ResNet50 models from the Zoo been done? I mean formal verification, even beyond the test for VGG19 that appears here: https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-zoo/src/test/java/org/deeplearning4j/zoo/TestImageNet.java

Although, something like that would be a good start. The reason I ask is that I tried using DL4J's GoogLeNet to get the same prediction numbers shown at this website, https://www.exaptive.com/blog/how-i-made-a-neural-network-web-application-in-an-hour , and the results were quite different. It's possible that this website can't be trusted, and I know that minor floating point errors can add up a bit, but I'm really uncertain about pre-processing, and whether I'm doing it correctly. That's why a nice test example would be nice.

The website I linked to gives the following predictions for a picture of a rabbit and kitten using GoogleNet:
72.1% hare
18.5% wood rabbit
4.5% lynx

My code uses the same procedure described on this site, https://deeplearning4j.org/build_vgg_webapp#example (my VGG16 predictions are identical by the way), and also in the VGG19 unit test linked to above, but with GoogLeNet, and I get the following results.

Using the VGG16Preprocessor:
Predictions for batch  :
	28.198389%, lynx
	14.360826%, tabby
	13.042391%, wood_rabbit
	11.612229%, hare
	9.108539%, tiger_cat

Without ANY preprocessing:
Predictions for batch  :
	26.153824%, fox_squirrel
	19.492746%, wood_rabbit
	7.596114%, lynx
	7.167132%, hare
	6.885763%, tabby

What is really frustrating about these results is that they seem to be reasonable, whether I use the preprocessor or not, but this is a single example image, and I can't even replicate another result that supposedly used GoogLeNet. I can't be certain I am using this correctly if I can't replicate an established result.

**Comment by tomthetrainer** [comment]
A quick search, not 100 % sure it is accurate shows. 

Googlenet pre-processing is mean subtraction. 
crop_size: 224
mean_value: 104
mean_value: 117
mean_value: 123

Here is an example of a pre-processor for VGG16, similar process, load the image, subtract the mean. 

https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/preprocessor/VGG16ImagePreProcessor.java

**Comment by schrum2** [comment]
That is reassuring, and it means that I can/should use the VGG16Preprocessor with GoogLeNet. Still, having some actual reference images along with specific expected prediction scores would be nice. I was previously using the VGG16Preprocessor, and my numbers still differed from those on this site: https://www.exaptive.com/blog/how-i-made-a-neural-network-web-application-in-an-hour

**Comment by saudet** [comment]
@schrum2 IIRC, GoogLeNet needs image channels to be in RGB order, not BGR, but some imported models do use BGR as well, for Caffe. (DL4J is also BGR by default as that's how OpenCV loads images.) Not sure which GoogLeNet model got imported in DL4J...

**Comment by schrum2** [comment]
Ok, got images loaded using NativeImageLoader loader = new NativeImageLoader(224, 224, 3, new ColorConversionTransform(org.bytedeco.javacpp.opencv_imgproc.COLOR_BGR2RGB));
		
However, now I can't use the VGG16Preprocessor, if that uses the BGR order, so I made my own GoogLeNetImagePreprocessor class that just switches the order of values in VGG_MEAN_OFFSET_BGR. However, this gives weird predictions. I'm still comparing against the site here: https://www.exaptive.com/blog/how-i-made-a-neural-network-web-application-in-an-hour

Using my new GoogLeNetImagePreprocessor:
Predictions for batch  :
	11.722349%, tabby
	9.900959%, Egyptian_cat
	7.218280%, triceratops
	7.051530%, lynx
	4.992188%, tiger_cat

Using no preprocessing at all:
Predictions for batch  :
	14.228871%, common_iguana
	10.925683%, triceratops
	8.812419%, American_alligator
	7.287676%, frilled_lizard
	6.306552%, African_crocodile

I feel like the predictions before were better, but who can tell how well the network will perform on any specific image? It is also possible that the site I'm referring to is wrong, or maybe there is something weird about the image I copied ... once again, this is why I would really like to have a verifiable set of numbers to compare to.

Without them, I can tweak and vary lots of things, and keep getting predictions, but it will be hard to trust them. When the network gives a bad label, is that just one of those bad examples that is hard to process, or is it because I configured something wrong? I just wish there was a similar level of comparison as exists with the VGG16 web app.

**Comment by schrum2** [comment]
More useful information. Tried matching the results from this site:
https://www.tensorflow.org/tutorials/image_recognition
There is a Panda image here. However, these scores are from Inception-v3, and maybe the GoogLeNet in DL4J is Inception-v1? In any case, I tried all possible combinations of using and not using different preprocessors and loading in RGB vs BGR and in the best case got a Panda label confidence of around 55% as opposed to the 88% reported on this webpage.

**Comment by schrum2** [comment]
I found an even better way to test how GoogLeNet should perform. This site:
http://cs.stanford.edu/people/karpathy/ilsvrc/
This site is meant to allow a human try the ImageNet challenge, but the reason it is useful is that there is a "Show google prediction" button which is referring to GoogLeNet. For any of the images, you can click the button and show the top labels GoogLeNet describes. Unfortunately, the site does not give the prediction confidence numbers.

However, I tested DL4J's GoogLeNet with and without preprocessing, and with the images loaded in both BGR and RGB, and I never got the exact same ordered sequence of image labels. I did frequently get the same top-level prediction, which I guess is good. In fact, I could get the same correct label sometimes regardless of the preprocessing or image color ordering, which I guess indicates that the network is quite robust to weird color changes. 

However, I still don't see why I can't get the exact same order of the exact same prediction labels. Either there is something wrong with the network weights (seems unlikely because they are directly copied from elsewhere), something wrong with the code that builds and runs the networks (possible I guess), or something wrong with how I am loading the images and sending them to the network (this is why I'm asking for help).

**Comment by saudet** [comment]
It's quite possible GoogLeNet isn't working properly, @crockpotveggies might have more information.

**Comment by crockpotveggies** [comment]
Specifically GoogLeNet, @raver119 and I have identified a couple issues there that we've been looking at outside of Github. However, I'm not aware of issues with ResNet50. ResNet50 was directly converted from the Keras applications repository, so issues with that particular model would be either preprocessing or import failure.

We used to test *all* models but this wasn't feasible, the bandwidth on our CI server was max'd out and would throw irrelevant errors.

I'm open to suggestions here.

**Comment by saudet** [comment]
We don't need to run the tests on the CI server, as long as there would be tests to run, that would be a good start I think. I tried adding some of the other models to the unit tests, and some of them were failing:
https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-zoo/src/test/java/org/deeplearning4j/zoo
So, just having those up there would be a good indications of their health, IMO. What do you think?

**Comment by crockpotveggies** [comment]
Yes I'm all for testing. So do we want to remove the @Test annotation on a
"special" test for this? I'll poke @huitseeker about running specific tests.

**Comment by saudet** [comment]
No, no need to remove anything. IMO, we should be adding more models to the current set of "smoke tests", because a lot of the models are failing even those (when they are put in the tests that is).

**Comment by schrum2** [comment]
What is the current state of GoogLeNet? Is it's correctness still uncertain?

**Comment by saudet** [comment]
Yes, still needs checking.

**Comment by AlexDBlack** [comment]
@crockpotveggies one option here: test *all* of them, but have it behind a profile that is disabled by default.
That way, they won't be run by CI on every build for example, but we can still run them manually, periodically and before releases.

**Comment by agibsonccc** [comment]
Closing — references a removed/rewritten subsystem (DataVec, NLP/Word2Vec, Model Zoo, Spark, UI, or Arbiter). These have been fundamentally rewritten or removed in the upcoming release. Please open a new issue if you experience similar problems with the new version.
